source/material_model/rheology/composite_visco_plastic.cc

```diff
@@ -82,38 +82,302 @@ namespace aspect
+                    viscoplastic_viscosity_guess_c = std::min(dislocation_creep->compute_viscosity(edot_ii, pressure, temperature, composition, phase_function_values, n_phase_transitions_per_composition), viscoplastic_viscosity_guess_c);
+                  }
+
+                if (use_peierls_creep)
+                  {
+                    peierls_creep_parameters.push_back(peierls_creep->compute_creep_parameters(composition));
+                    viscoplastic_viscosity_guess_c = std::min(peierls_creep->compute_approximate_viscosity(edot_ii, pressure, temperature, composition), viscoplastic_viscosity_guess_c);
+                  }
+
+                if (use_drucker_prager)
+                  {
+                    Rheology::DruckerPragerParameters drucker_prager_parameters_c = drucker_prager->compute_drucker_prager_parameters(composition, phase_function_values, n_phase_transitions_per_composition);
+                    drucker_prager_parameters.push_back(drucker_prager_parameters_c);
+                    viscoplastic_viscosity_guess_c = std::min(drucker_prager->compute_viscosity(drucker_prager_parameters_c.cohesion,
+                                                                                                drucker_prager_parameters_c.angle_internal_friction, pressure, edot_ii, drucker_prager_parameters_c.max_yield_stress),
+                                                              viscoplastic_viscosity_guess_c);
+                  }
+                inverse_viscoplastic_viscosity_guess += volume_fraction / viscoplastic_viscosity_guess_c;
               }
             else
               {
                 total_volume_fraction -= volume_fractions[composition];
               }
+          }
+
+        const double viscoplastic_viscosity_guess = 1. / inverse_viscoplastic_viscosity_guess;
+        double viscoplastic_stress = 2. * viscoplastic_viscosity_guess * edot_ii / total_volume_fraction;
+        double log_viscoplastic_stress = std::log(viscoplastic_stress);
+
+        // 2) Calculate the log strain rate and first derivative with respect to
+        // log stress for each element using the guessed creep stress.
+        std::vector<std::array<std::pair<double, double>, 4>> log_edot_and_deriv(active_compositions.size());
 
-            viscosity /= total_volume_fraction;
-            for (unsigned int j=0; j < 5; ++j)
-              partial_strain_rates[j] /= total_volume_fraction;
+        for (unsigned int i = 0; i < active_compositions.size(); ++i)
+          {
+            const double log_volume_fraction = std::log(volume_fractions[active_compositions[i]]);
+
+            if (use_diffusion_creep)
+              {
+                log_edot_and_deriv[i][0] = diffusion_creep->compute_log_strain_rate_and_derivative(log_viscoplastic_stress, pressure, temperature, grain_size, diffusion_creep_parameters[i]);
+                log_edot_and_deriv[i][0].first += log_volume_fraction;
```

Comment:
why do you add the log_volume_fraction to the log strain rate? is it because the addition in log-space is the same as the multiplication in without the log? please add a comment to explain.